### PR TITLE
db/hints: Make commitlog instances use the commitlog IO scheduling group

### PR DESCRIPTION
**Pull request by dawmd**
Before these changes, we didn't specify which I/O scheduling group commitlog instances in hinted handoff should use. In this PR, we set it explicitly to the commitlog scheduling group. The rationale for this choice is the fact we don't want to cause a bottleneck on the write path -- if hints are written too slowly, new incoming mutations (NOT hints) might be rejected due to a too high number of hints currently being written to disk; see `storage_proxy::create_write_response_handler_helper()` for more context.

Fixes scylladb/scylladb#18654

### CODE REVIEW

**Comment by dawmd** [comment]
The CI failure seems unrelated to my changes. Rerunning it

**Comment by piodul** [comment]
> The CI failure seems unrelated to my changes. Rerunning it

Perhaps, but please check whether it's an existing issue or it's something new. If it's something new - create a new issue and attach information about the failure and the relevant logs.

**Comment by piodul** [comment]
> ❌ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9543/)

> ERROR: Build x86_64 failed with error: script returned exit code 137

According to The Internet, this means an OOM in the context of Docker. Restarting.

**Comment by piodul** [comment]
> Before these changes, we didn't specify which I/O scheduling group commitlog instances in hinted handoff should use.

I think it's important to add some more context here:

- In the past, scheduling groups and io priority classes were separate, and both main commitlog and the hints commitlogs used to use the same io priority class
- In https://github.com/scylladb/scylladb/pull/14005 we started to set the scheduling group explicitly as a preparation for unifying scheduling groups and io priority classes, however it does not set the SG for hints commitlogs
- Eventually, seastar unified SGs and IO priority classes and we started observing that hints run in the main IO priority class, which does not sound like the right place

File: service/storage_proxy.cc [hunk 1]
```diff
@@ -3897,7 +3895,8 @@ template<typename Range>
 bool storage_proxy::cannot_hint(const Range& targets, db::write_type type) const {
     // if hints are disabled we "can always hint" since there's going to be no hint generated in this case
     return hints_enabled(type) &&
-            _hints_manager.started() &&
+            _hints_manager.started() && // If the manager hasn't started yet, no mutation will be performed to another node.
+                                        // No hint will need to be stored.
```
Comment (piodul) [inline service/storage_proxy.cc:3899]: I'm all for incremental improvements and cleanups, but this one is completely unrelated to the topic of this PR. Could you send it as a separate pull request?

**Comment by xemul** [comment]
Technically the change is correct, but whether regular commitlog sched group should be used is good question. We have 'schema_commitlog' one for schema changes, maybe we need dedicated one for hints? Or maybe just maintenace one?

**Comment by piodul** [comment]
@xemul hints can be generated in large volumes, potentially on each write. An extreme example would be: imagine a 3-node cluster, RF=3 keyspace, one node is down. Each coordinated write will have to write to the regular commitlog (because coordinator are always replicas here) and each write will also generate a hint because one of the replicas is down.

Moreover, if a node doesn't keep up with writing hints, it will reject writes until it is able to store more hints.

I don't think a maintenance/lower priority group will be a good fit.

**Comment by xemul** [comment]
Agree, maintenance group is bad choice for sure.

**Comment by dawmd** [comment]
I personally think hinted handoff should have a "named" scheduling group anyway, even if it ends up the same as some other one, e.g. commitlog's one.

**Comment by kostja** [comment]
For materialized views, hint writes are as important as commit log writes. The write bandwidth created by hints is similar - it's O(commit log writes). If there is a separate scheduling group, then the question is how it should compete for bandwidth with the commit log and compaction. I think we should not create one unless there is some benchmarking evidence to do it. Which begs the question, @dawmd how did you check the pull request addresses the issue?

**Comment by dawmd** [comment]
> Which begs the question, @dawmd how did you check the pull request addresses the issue?

I didn't and I see why that's a blocker. I might be wrong, but I think we should measure the performance impact using an actual multi-node cluster to get accurate results, e.g. on AWS, where we stress it with both view updates and regular hints. I can test it locally and collect metrics first, though.

**Comment by piodul** [comment]
> I didn't and I see why that's a blocker. I might be wrong, but I think we should measure the performance impact using an actual multi-node cluster to get accurate results, e.g. on AWS, where we stress it with both view updates and regular hints. I can test it locally and collect metrics first, though.

The first thing that you should verify is that, after your changes, there is little to no traffic in the "main" I/O group.

As for the benchmark - I don't think that setting up a cluster in AWS is necessary. You can run a benchmark on your PC or on some of the computers in our office. You can try overloading the disk by simulating a workload that sends large amounts of data (e.g. via cql-stress). You should get the "too many in-flight hints" error if hints won't keep up with writing to disk. I suspect that, after your changes, you should get much less of these errors for the same workload.

**Comment by dawmd** [comment]
Version 3:
* Rebased on top of `master`.

**Comment by dawmd** [comment]
We wanted to verify the hypothesis that after these changes, Scylla will experience fewer `too_many_in_flight_hints` errors. Because throwing that error prevents the database from processing a mutation, a side effect should be that we'll also be able to perform more writes and save more hints.

---

I tested the solution using the following test scenario:

1. Create a cluster consisting of three nodes.
2. Run cassandra-stress for some time as a preparatory step.
3. Run cassandra-stress for 90 minutes, sending big mutations to the nodes. After 30 minutes from its start, choose one of the nodes at random and stop it. After another 30 minutes, turn on that node again. Continue running C-S for the last 30 minutes.

---

Hardware used in the test:

1. Loaders: 5 instances of `c5.2xlarge`.
2. Monitor: 1 instance of `t3.large`.
3. Nodes: 3 instances of `i3.2xlarge`.

The cassandra-stress commands used in the test:

1. Preparatory phase:

```bash
cassandra-stress write no-warmup cl=ALL n=62500000 -schema 'replication(strategy=NetworkTopologyStrategy,replication_factor=3)' -mode cql3 native -rate threads=200 -col 'size=FIXED(128) n=FIXED(16)' -pop seq=[...] -errors ignore
```

where `[...]` stands for consecutive ranges.

2. The stress phase:

```bash
cassandra-stress write no-warmup cl=QUORUM duration=600m -schema 'replication(strategy=NetworkTopologyStrategy,replication_factor=3)' -mode cql3 native -rate 'threads=200' -col 'size=FIXED(2048) n=FIXED(16)' -pop 'dist=gauss(1..250000000,125000000,12500000)' -errors ignore
```

The rationale for choosing 16 columns and a relatively big size was to trigger `too_many_in_flight_hints` errors; hinted handoff doesn't allow for more than 10 MB of concurrent hints being written to disk at a given time. When that threshold is achieved, a write response handler throws said error.

3. Scylla arguments:

```bash
--blocked-reactor-notify-ms 5 --abort-on-lsa-bad-alloc 1 --abort-on-seastar-bad-alloc --abort-on-internal-error 1 --abort-on-ebadf 1
```

---

### Results overview
TODO


---

<details>
<summary>Graphs</summary>

### Total size of hints in flight per instance
#### master
total_size_of_hints_in_flight_master

#### this branch
total_size_of_hints_in_flight_this_branch

### Total count of written hints per instance
#### master
scylla_hints_manager_written_master

#### this branch
scylla_hints_manager_written_this_branch

### Hints written per minute and per instance
#### master
hints_written_per_minute_and_per_instance_master

#### this branch
hints_written_per_minute_and_per_instance_this_branch

### Errors due to too many hints in flight per instance (failed writes)
#### master
scylla_storage_proxy_coordinator_writes_failed_due_to_too_many_in_flight_hints_master

#### this branch
scylla_storage_proxy_coordinator_writes_failed_due_to_too_many_in_flight_hints_this_branch

### Write timeouts/second per instance
#### master
image

#### this branch
image

### Write unavailable/second per instance
#### master
image

#### this branch
image

### Latencies
#### master
image

#### this branch
image

### Scheduling groups information
#### master
image

#### this branch
image

### Miscellaneous
#### master
image

#### this branch
image

### C++ exceptions per instance
#### master
image

#### this branch
image

</details>

---

SCT runs IDs for restoring the monitor and other data:
* master: `aed4e844-0b4c-4dd6-9948-4dad431e07da`
* this branch: `3e4dfe95-145d-4e66-b988-47fa0c2e1256`

**Comment by mykaul** [comment]
VERY nice report @dawmd  - thank you!
I did not understand what was the tradeoff though - the fact we move a scheduling group is not magic - it's on the expense of something else. 
The P99 is interesting - 2 secs is quite unacceptable - so perhaps it worked well when the cluster was fully stressed?

**Comment by dawmd** [comment]
> VERY nice report @dawmd - thank you!

Thank you!

> I did not understand what was the tradeoff though - the fact we move a scheduling group is not magic - it's on the expense of something else.

I suspect that this particular test scenario couldn't faithfully reflect it. Because the down node is a replica for every write we perform and because a write cannot succeed if we can't save a hint for *any* of the replicas (cf. `storage_proxy::create_write_response_handler_helper()` and `storage_proxy::cannot_hint()`), all of the writes cassandra-stress sends depends on whether we can save a hint to that down node. Since we limit the concurrency of that (we can't be writing more 10 MB of hints per shard at a time), that in turn depends on how fast we perform the I/O operations. If we do it too slow – a write will fail even if the consistency level could be achieved without that down replica.

That's why the "competition" between the branches boiled down to the efficiency of hinted handoff – and that's exactly what we wanted to verify. We could perform another, perhaps more realistic test with smaller mutations and maybe more nodes – some mutations would depend on whether we can save a hint for a down node, but some would not.

What should happen in general, to my understanding, is we'd start failing as many mutations at an early stage of performing a write (because of fewer `too_many_in_flight_hints` errors preventing performing it in the first place), but the price would be a bit slower writes to the "main" commitlog (the one corresponding to the writes to memtables), compactions, etc.

> The P99 is interesting - 2 secs is quite unacceptable - so perhaps it worked well when the cluster was fully stressed?

Maybe. But frankly speaking, I'm a bit surprised with the result. I would've thought the cluster could sustain more mutations if not for the down replica, so 2 seconds seems quite high.

File: db/hints/internal/hint_endpoint_manager.cc [hunk 2]
```diff
@@ -196,6 +196,7 @@ future<db::commitlog> hint_endpoint_manager::add_store() noexcept {
         return io_check([name = _hints_dir.c_str()] { return recursive_touch_directory(name); }).then([this] () {
             commitlog::config cfg;
 
+            cfg.sched_group = _shard_manager.local_db().commitlog()->active_config().sched_group;
```
Comment (xemul) [inline db/hints/internal/hint_endpoint_manager.cc:199]: Am I right that before this patch hints commitlog was using default sched group?

Comment (dawmd) [reply db/hints/internal/hint_endpoint_manager.cc:199]: Yes

**Review by piodul — APPROVED** [review]
I'll merge it:

- The benchmark showed positive effect on the number of returned overloaded errors - we are less frequently bottlenecked by hints and the bottleneck is moved to other parts of the system, for which we wait for completion instead of rejecting immediately,
- The default scheduling group is a bad choice for this kind of workload in general,
- The `commitlog` scheduling group was used in the past, we are bringing old behavior which was accidentally changed.

**Comment by piodul** [comment]
Queued

**Comment by avikivity** [comment]
> Makes sense to me - just one nitpick, pre-approving.
> 
> AFAIK the hints commitlog now runs in the `main` scheduling group by accident which does not sound like the right scheduling group for it. The "commitlog" group used to be used previously for this, so there is a precedent.
> 
> @avikivity @xemul could you also review this small change? What do you think about the scheduling group choice for writing hints to disk?

This is not a small change!

But I believe it's correct. A a hint is an extension of the normal write path. So it can use the same scheduling group.

Hint writes will compete with regular writes, but that's a good thing. We do want to slow down regular writes if hints take up too much bandwidth.
